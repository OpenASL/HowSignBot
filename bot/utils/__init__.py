import difflib
import re
from typing import Optional
from typing import Sequence
from typing import Tuple

_spoiler_pattern = re.compile(r"\s*\|\|\s*(.*)\s*\|\|\s*")
_quotes_pattern = re.compile(r"[\"“](.*?)[\"”]")


def get_spoiler_text(val: str) -> Optional[str]:
    """Return value within spoiler text if it exists, else return `None`."""
    match = _spoiler_pattern.match(val)
    if match:
        return match.groups()[0]
    return None


def get_and_strip_quoted_text(val: str) -> Tuple[str, Optional[str]]:
    """Return `val` with quoted text removed as well as as the quoted text."""
    match = _quotes_pattern.search(val)
    if match:
        stripped = _quotes_pattern.sub("", val).strip()
        quoted = match.groups()[0]
        return stripped, quoted
    return val, None


def truncate(s: str, max_len: int, *, trailing: str = "…"):
    if len(s) > max_len:
        return f"{s[:max_len]}{trailing}"
    return s


def get_close_matches(word: str, possibilities: Sequence[str]) -> Sequence[str]:
    return difflib.get_close_matches(word, possibilities, n=1, cutoff=0.5)


def did_you_mean(word: str, possibilities: Sequence[str]) -> Optional[str]:
    try:
        return get_close_matches(word, possibilities)[0]
    except IndexError:
        return None
