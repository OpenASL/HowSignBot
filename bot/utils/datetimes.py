import datetime as dt
from typing import Optional
from typing import Tuple

import dateparser
import pytz
from pytz.tzinfo import StaticTzInfo

import pytz_informal


def utcnow() -> dt.datetime:
    return dt.datetime.now(dt.timezone.utc)


PACIFIC = pytz.timezone("America/Los_Angeles")
MOUNTAIN = pytz.timezone("America/Denver")
CENTRAL = pytz.timezone("America/Chicago")
EASTERN = pytz.timezone("America/New_York")

# EDT and PDT change to EST and PST during the winter
# Show the current name in docs
EASTERN_CURRENT_NAME = utcnow().astimezone(EASTERN).strftime("%Z")
PACIFIC_CURRENT_NAME = utcnow().astimezone(PACIFIC).strftime("%Z")

# Timezone is omitted because it is computed using tzinfo.tzname
TIME_FORMAT = "%-I:%M %p "
TIME_FORMAT_NO_MINUTES = "%-I %p "


def normalize_timezone(dtime: dt.datetime) -> dt.datetime:
    """Normalizes informal N. American timezones ("EST", "PST") to
    the IANA timezones ("America/Los_Angeles", "America/New_York")
    """
    tzinfo = dtime.tzinfo
    assert tzinfo is not None
    naive = dtime.replace(tzinfo=None)
    tzname = tzinfo.tzname(naive)
    assert tzname is not None
    tzone = pytz_informal.timezone(tzname)
    return tzone.localize(naive)


class NoTimeZoneError(ValueError):
    pass


def parse_human_readable_datetime(
    dstr: str,
    settings: Optional[dict] = None,
    user_timezone: Optional[pytz.BaseTzInfo] = None,
    # By default, use Pacific time if timezone can't be parsed
    fallback_timezone: Optional[pytz.BaseTzInfo] = PACIFIC,
) -> Tuple[Optional[dt.datetime], Optional[dt.tzinfo]]:
    parsed = dateparser.parse(dstr, settings=settings)
    if parsed is None:
        return None, None
    if not parsed.tzinfo:
        if user_timezone is not None:
            parsed = user_timezone.localize(parsed)
        else:
            if not fallback_timezone:
                raise NoTimeZoneError(f"Time zone could not be parsed from {dstr}.")
            parsed = fallback_timezone.localize(parsed)
    parsed = normalize_timezone(parsed)
    used_timezone = parsed.tzinfo
    return parsed.astimezone(dt.timezone.utc), used_timezone


def display_timezone(tzinfo: StaticTzInfo, dtime: dt.datetime) -> str:
    # Pass is_dst False to handle ambiguous datetimes
    # NOTE: America/Los_Angeles will still display correctly as PDT after DST ends
    ret = tzinfo.tzname(dtime.replace(tzinfo=None), is_dst=False)
    assert ret is not None
    return ret


def display_time(dtime: dt.datetime, time_format: str, tzinfo: pytz.BaseTzInfo) -> str:
    return dtime.astimezone(tzinfo).strftime(time_format) + display_timezone(
        tzinfo, dtime
    )


def format_multi_time(dtime: dt.datetime) -> str:
    time_format = TIME_FORMAT if dtime.minute != 0 else TIME_FORMAT_NO_MINUTES
    pacific_dstr = display_time(dtime, time_format, tzinfo=PACIFIC)
    mountain_dstr = display_time(dtime, time_format, tzinfo=MOUNTAIN)
    central_dstr = display_time(dtime, time_format, tzinfo=CENTRAL)
    eastern_dstr = display_time(dtime, time_format, tzinfo=EASTERN)
    return " / ".join((pacific_dstr, mountain_dstr, central_dstr, eastern_dstr))
