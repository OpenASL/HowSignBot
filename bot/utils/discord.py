import discord


def display_name(user: discord.User | discord.Member) -> str:
    return getattr(user, "nick", None) or user.name
