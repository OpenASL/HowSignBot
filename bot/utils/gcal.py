import datetime as dt
from urllib.parse import urlencode


def create_gcal_url(
    text,
    start: dt.datetime,
    end: dt.datetime | None = None,
    description: str | None = None,
):
    dt_format = "%Y%m%dT%H%M%SZ"
    base_url = "http://www.google.com/calendar/event"
    end = end or start + dt.timedelta(hours=1)
    params = {
        "action": "TEMPLATE",
        "text": text,
        "dates": f"{start.strftime(dt_format)}/{end.strftime(dt_format)}",
    }
    if description:
        params["details"] = description
    return "?".join((base_url, urlencode(params)))
