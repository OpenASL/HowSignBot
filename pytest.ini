[pytest]
filterwarnings =
    ignore:.*@coroutine.*:DeprecationWarning
