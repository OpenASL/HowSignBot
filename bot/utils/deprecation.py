from __future__ import annotations

from contextlib import suppress

import disnake
from disnake.ext.commands import Context

from bot import settings

COMMAND_PREFIX = settings.COMMAND_PREFIX

TEMPLATE = """ðŸ’¡ Heads up: the `{COMMAND_PREFIX}{ctx.invoked_with}` command is being phased out. Use `{substitute}` instead.
Before:
```
{before_example}
```
After:
```
{after_example}
```
"""


async def send_deprecation_notice(
    ctx: Context, *, substitute: str, before_example: str, after_example: str
):
    if settings.SEND_DEPRECATION_MESSAGES is False:
        return
    content = TEMPLATE.format(
        COMMAND_PREFIX=COMMAND_PREFIX,
        ctx=ctx,
        substitute=substitute,
        before_example=before_example,
        after_example=after_example,
    )
    with suppress(disnake.errors.Forbidden):  # user may not allow DMs from bot
        await ctx.author.send(content=content)
