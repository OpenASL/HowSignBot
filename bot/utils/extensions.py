# From https://github.com/python-disnake/bot/blob/master/bot/utils/extensions.py
import importlib
import inspect
import pkgutil
from typing import Iterator
from typing import NoReturn

from bot import exts


def unqualify(name: str) -> str:
    """Return an unqualified name given a qualified module/package `name`."""
    return name.rsplit(".", maxsplit=1)[-1]


def walk_extensions() -> Iterator[str]:
    """Yield extension names from the bot.exts subpackage."""

    def on_error(name: str) -> NoReturn:
        raise ImportError(name=name)  # pragma: no cover

    for module in pkgutil.walk_packages(
        exts.__path__, f"{exts.__name__}.", onerror=on_error  # type: ignore
    ):
        if unqualify(module.name).startswith("_"):
            # Ignore module/package names starting with an underscore.
            continue

        if module.ispkg:
            imported = importlib.import_module(module.name)
            if not inspect.isfunction(getattr(imported, "setup", None)):
                # If it lacks a setup function, it's not an extension.
                continue

        yield module.name
