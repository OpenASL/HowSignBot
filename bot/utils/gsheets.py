import gspread
from google.auth.crypt._python_rsa import RSASigner
from google.oauth2.service_account import Credentials

from bot.settings import GOOGLE_CLIENT_EMAIL
from bot.settings import GOOGLE_PRIVATE_KEY
from bot.settings import GOOGLE_PRIVATE_KEY_ID
from bot.settings import GOOGLE_PROJECT_ID
from bot.settings import GOOGLE_TOKEN_URI


def get_gsheet_client():
    signer = RSASigner.from_string(key=GOOGLE_PRIVATE_KEY, key_id=GOOGLE_PRIVATE_KEY_ID)
    credentials = Credentials(
        signer=signer,
        service_account_email=GOOGLE_CLIENT_EMAIL,
        token_uri=GOOGLE_TOKEN_URI,
        scopes=gspread.auth.DEFAULT_SCOPES,
        project_id=GOOGLE_PROJECT_ID,
    )
    return gspread.authorize(credentials)
