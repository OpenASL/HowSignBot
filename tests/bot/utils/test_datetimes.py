import datetime as dt

from freezegun import freeze_time
import pytest
import pytz

from bot.utils.datetimes import parse_human_readable_datetime, display_timezone


@pytest.mark.parametrize(
    "value",
    (
        "today 8:25pm edt",
        "today 2pm edt",
        "today at 2pm edt",
        "tomorrow 2pm pdt",
        "on friday at 2pm cst",
        "9/25 2:30 pm cdt",
    ),
)
@freeze_time("2020-09-25 14:00:00")
def test_parse_human_readable_datetime(value, snapshot):
    dtime, _ = parse_human_readable_datetime(value)
    assert dtime.tzinfo == dt.timezone.utc
    assert dtime.isoformat() == snapshot


@pytest.mark.parametrize(
    ("value", "expected"),
    (
        (pytz.timezone("America/New_York"), "EDT"),
        (pytz.timezone("America/Los_Angeles"), "PDT"),
    ),
)
def test_display_timezone(value, expected):
    dtime = dt.datetime(2020, 9, 25, tzinfo=dt.timezone.utc)
    assert display_timezone(value, dtime) == expected
