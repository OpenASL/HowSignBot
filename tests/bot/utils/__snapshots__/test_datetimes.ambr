# name: test_parse_human_readable_datetime[9/25 2:30 pm cdt]
  '2020-09-25T19:30:00+00:00'
---
# name: test_parse_human_readable_datetime[on friday at 2pm cst]
  '2020-09-25T19:00:00+00:00'
---
# name: test_parse_human_readable_datetime[today 2pm edt]
  '2020-09-25T18:00:00+00:00'
---
# name: test_parse_human_readable_datetime[today 8:25pm edt]
  '2020-09-26T00:25:00+00:00'
---
# name: test_parse_human_readable_datetime[today at 2pm edt]
  '2020-09-25T18:00:00+00:00'
---
# name: test_parse_human_readable_datetime[tomorrow 2pm pdt]
  '2020-09-26T21:00:00+00:00'
---
